### DISCUSSION TITLE
Math texts with superscript and subscript in adjacent lines overlap with each other in terms lists

### DISCUSSION THREAD

**Discussion by zhangfand** [typst/typst#2620] [General]
### Description

Here is a minimal reproducible example of the issue:

```
/ $n_y$: output size (or number of classes)
/ $n_h^[l]$: number of hidden units of the layer $l$.
```

Here is the rendered pdf. 
<img width="365" alt="Screenshot 2023-11-07 at 10 04 34 PM" src="https://github.com/typst/typst/assets/6268519/21ebc3bb-a2a8-4c01-bb3c-9af4ca4376c4">

### Operating system

Web app

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enivex** [comment]
This is not a bug. The bounding box of inline math purposely does not match its content (up to a point), because it would quickly affect line spacing. you have three options:

- Use block math instead of inline
- Increase spacing between terms
- Rewrite so that the tall things aren't exactly on top of each other.

**Comment by laurmaedje** [comment]
I think it's kind of a mix. Equations shouldn't affect line spacing when possible, but sometimes they _have_ to and Typst doesn't do that yet in an ideal way. There's an existing issue here though: https://github.com/typst/typst/issues/1028

**Comment by zhangfand** [comment]
The way LaTex handles this issue is it will increases the line spacing. See this question for an example. https://tex.stackexchange.com/questions/116479/how-to-force-normal-line-spacing-when-inline-math-in-text

Whether that's a sane design for typst to adopt probably needs to be discussed.

**Comment by vsolon7** [comment]
I was going to make a new post, but then I found this, so I figure I'd leave a comment with more exposition and a (hopefully not terrible) idea. Here's an image for reference. I've reduced the line spacing to `1pt` with `#show par: set block(spacing: 1pt)` and drawn boxes around all of the content (and adjusted the inline math box to be where it should be, not just centered based on the box size):
![image](https://github.com/typst/typst/assets/671323/48685fdf-7528-4e60-ba48-01e0e2c17026)

Based on the boxes, whose heights are set by the height of the content they contain, it seems that typst creates some kind of fake height for inline math (as well as font, but there I know the box height is determined by the parameters `top-edge` and `bottom-edge` of the `text` function), and then takes this fake height as the true height of the content. This fake height is created with the default line spacing in mind, I assume, as the same content with default line spacing gives
![image](https://github.com/typst/typst/assets/671323/48b5577b-2322-4a6d-bab4-5602e26f440f)
which is much better. You'd never know something was weird --- the math bleeds a little into the line spacing, but doesn't impact the actual text, in a similar way to TeX. 

I can see why this is somewhat better than simply using the real height of the inline math content, as this way makes it so inline math doesn't create way too much space. But with no way to keep track of the true height of the content, you have the opposite issue, where it doesn't create enough space as it doesn't think the text is overlapping.

Wouldn't the optimal way be to keep track of the *true* height of all letters and inline math content, place them on the baseline of the line, then check to see if the height of the content above the baseline is larger than the line height + line spacing (or probably "distance to the glyphs above the content")? That way, if it is, you add more line spacing above the current line so that line height + line spacing is larger than the height of the content above the baseline (and similar for below the baseline).

Admittedly, I'm not entirely sure how typst and TeX work, but it seems like TeX does something that is essentially functionally equivalent to this based on examples I've tested. I can't see any better way to do things, as leaving the inline math bleeding into the line above/below is probably bad. In my opinion, this is one of the major things that TeX has an edge over typst on. It has more data about each glyph, so the typesetting doesn't run into these issues (or at least not nearly as much). I can't think of why you would ONLY want to know the "fake" height of the content.

**Reply by laurmaedje** [reply]
The current implementation was mostly a quick hack to get things working reasonably in the defaults. Something smarter like you describe should definitely be implemented.
